src/decoder/mod.rs

```diff
@@ -359,95 +359,76 @@ impl Wrapping for i64 {
-        }
-        DecodingBuffer::U32(buf) => {
-            rev_hpredict_nsamp(buf, size, img_width, samples)?;
-        }
-        DecodingBuffer::U64(buf) => {
-            rev_hpredict_nsamp(buf, size, img_width, samples)?;
-        }
-        DecodingBuffer::F32(_buf) => {
-            // FIXME: check how this is defined.
-            // See issue #89.
-            // rev_hpredict_nsamp(buf, size, img_width,samples)?;
-            return Err(TiffError::UnsupportedError(
-                TiffUnsupportedError::HorizontalPredictor(color_type),
-            ));
-        }
-        DecodingBuffer::F64(_buf) => {
-            //FIXME: check how this is defined.
-            // See issue #89.
-            // rev_hpredict_nsamp(buf, size, img_width,samples)?;
-            return Err(TiffError::UnsupportedError(
-                TiffUnsupportedError::HorizontalPredictor(color_type),
-            ));
-        }
-        DecodingBuffer::I8(buf) => {
-            rev_hpredict_nsamp(buf, size, img_width, samples)?;
-        }
-        DecodingBuffer::I16(buf) => {
-            rev_hpredict_nsamp(buf, size, img_width, samples)?;
+pub fn fp_predict_f64(input: &mut [u8], output: &mut [f64], samples: usize) {
+    rev_hpredict_nsamp(input, samples);
+    for i in 0..output.len() {
+        // TODO: use f64::from_be_bytes() when we can (version 1.40)
+        output[i] = f64::from_bits(u64::from_be_bytes([
+            input[input.len() / 8 * 0 + i],
+            input[input.len() / 8 * 1 + i],
+            input[input.len() / 8 * 2 + i],
+            input[input.len() / 8 * 3 + i],
+            input[input.len() / 8 * 4 + i],
+            input[input.len() / 8 * 5 + i],
+            input[input.len() / 8 * 6 + i],
+            input[input.len() / 8 * 7 + i],
+        ]));
+    }
+}
+
+fn fix_endianness_and_predict(
+    mut image: DecodingBuffer,
+    samples: usize,
+    byte_order: ByteOrder,
+    predictor: Predictor,
+) {
+    match predictor {
+        Predictor::None => {
+            fix_endianness(&mut image, byte_order);
         }
-        DecodingBuffer::I32(buf) => {
-            rev_hpredict_nsamp(buf, size, img_width, samples)?;
+        Predictor::Horizontal => {
+            fix_endianness(&mut image, byte_order);
+            match image {
+                DecodingBuffer::U8(buf) => rev_hpredict_nsamp(buf, samples),
+                DecodingBuffer::U16(buf) => rev_hpredict_nsamp(buf, samples),
+                DecodingBuffer::U32(buf) => rev_hpredict_nsamp(buf, samples),
+                DecodingBuffer::U64(buf) => rev_hpredict_nsamp(buf, samples),
+                DecodingBuffer::I8(buf) => rev_hpredict_nsamp(buf, samples),
+                DecodingBuffer::I16(buf) => rev_hpredict_nsamp(buf, samples),
+                DecodingBuffer::I32(buf) => rev_hpredict_nsamp(buf, samples),
+                DecodingBuffer::I64(buf) => rev_hpredict_nsamp(buf, samples),
+                DecodingBuffer::F32(_) | DecodingBuffer::F64(_) => unreachable!(),
```

Comment:
I'm very slightly uneasy about these assertions. These are strongly coupled with the accepted combinations in the decoder and we're one misstep in the check of decoder/image.rs:400 away from hitting this path. Can we avoid this coupling-at-adistance without too much overhead? Or may we just do a noop here, accepting a potentially wrong result in favor of stability.

Reply:
There's no especially elegant option here. Reporting the user-facing error here would require passing more context to this function that otherwise wouldn't be needed. And failing silently has its own downsides. The nice thing about panicking (especially with the added error messages) is that it is clear to a fuzzer/end users that there is a bug in the library.

However, I don't feel very strongly about this, so if there's a different option you'd prefer I'd be happy to change it.

Reply:
Just wondering here, with a message this seems good enough for the moment. Maybe we'll see a way to revisit this, and as you say, will have fuzzers. Although the fuzzer coverage very much depends on which code we're fuzzing because the edge coverage is not sufficient for this: we don't invoke the code reading into a pre-allocated buffer afaik.